main/src/ca/uwaterloo/flix/language/phase/unification/Zhegalkin.scala

```diff
@@ -0,0 +1,171 @@
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula.{Compl, Cst, ElemSet, Inter, Union, Var}
+import ca.uwaterloo.flix.util.CofiniteIntSet
+
+import scala.collection.immutable.SortedSet
+
+object Zhegalkin {
+
+  /** Represents a variable. */
+  case class ZhegalkinVar(v: Int) extends Ordered[ZhegalkinVar] {
+    override def toString: String = s"x$v"
+
+    override def compare(that: ZhegalkinVar): Int = this.v.compare(that.v)
+  }
+
+  /** Represents a set Zhegalkin constant (i.e. a set or co-set) */
+  case class ZhegalkinConstant(s: CofiniteIntSet) {
+    def compl: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.complement(s))
+    def union(that: ZhegalkinConstant): ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.union(s, that.s))
+    def inter(that: ZhegalkinConstant): ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.intersection(s, that.s))
+
+    override def toString: String = {
+      if (s.isEmpty) "Ø"
+      else if (s.isUniverse) "T"
+      else s match {
+        case CofiniteIntSet.Set(xs) => s"{${xs.mkString(", ")}}"
+        case CofiniteIntSet.Compl(xs) => s"~{${xs.mkString(", ")}}"
+      }
+    }
+  }
+
+  private val ZEmpty: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.empty)
+  private val ZUniverse: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.universe)
+
+  /** Represents a Zhegalkin term: c n x1 n x2 n ... n xn */
+  case class ZhegalkinTerm(cst: ZhegalkinConstant, vars: SortedSet[ZhegalkinVar]) {
+    override def toString: String =
+      if (vars.isEmpty)
+        cst.toString
+      else
+        s"$cst n ${vars.mkString(" n ")}"
+  }
+
+  /** Represents a Zhegalkin expr: c ? t1 ? t2 ? ... ? tn */
+  case class ZhegalkinExpr(cst: ZhegalkinConstant, terms: List[ZhegalkinTerm]) {
+    override def toString: String =
+      if (terms.isEmpty)
+        cst.toString
+      else
+        s"$cst ? ${terms.map(t => s"($t)").mkString(" ? ")}"
+  }
+
+  /**
+    * A smart constructor to filter empty intersections.
+    */
+  private def mkZhegalkinExpr(cst: ZhegalkinConstant, terms: List[ZhegalkinTerm]): ZhegalkinExpr =
+    ZhegalkinExpr(cst, terms.filter(t => !t.cst.s.isEmpty))
+
+  /** Returns the xor of the two Zhegalkin constants */
+  private def mkXor(c1: ZhegalkinConstant, c2: ZhegalkinConstant): ZhegalkinConstant = {
+    // a ? b = (a ? b) - (a n b) = (a ? b) n ~(a n b)
+    c1.union(c2).inter(c1.inter(c2).compl)
+  }
+
+  /** Returns the xor of the two Zhegalkin expressions. */
+  private def mkXor(z1: ZhegalkinExpr, z2: ZhegalkinExpr): ZhegalkinExpr = (z1, z2) match {
+    case (ZhegalkinExpr(c1, ts1), ZhegalkinExpr(c2, ts2)) =>
+      val c = mkXor(c1, c2)
+      val ts = (ts1 ++ ts2).groupBy(identity).collect { case (k, v) if v.size % 2 != 0 => k }.toList
```

Comment:
What's the idea here?

Reply:
It should be documented-- but so should a lot of the code.

Its the magic source, namely that for `a xor b xor a xor a` you only have `a` if it occurs an odd number of times.